x-pack/filebeat/input/awss3/states.go

```diff
@@ -0,0 +1,334 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package awss3
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/elastic/beats/v7/libbeat/statestore"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+// storedOp keeps track of pending updates that are not written to the persistent store yet.
+// Update operations are ordered. The input manager guarantees that only one
+// input can create update operation for a source, such that new input
+// instances can add update operations to be executed after already pending
+// update operations from older inputs instances that have been shutdown.
+type storedOp struct {
+	states  *States
+	store   *statestore.Store
+	lockMap *sync.Map
+}
+
+func newStoredOp(states *States, store *statestore.Store, lockMap *sync.Map) *storedOp {
+	return &storedOp{
+		states:  states,
+		store:   store,
+		lockMap: lockMap,
+	}
+}
+
+func (op *storedOp) execute(info s3Info) {
+	if op == nil {
+		return
+	}
+
+	id := info.name + info.key
+	state := op.states.FindPreviousByID(id)
+	if !state.IsEmpty() {
+		state.MarkAsStored()
+		op.states.Update(state, info.listingID)
+		op.states.writeStates(op.store)
+	}
+
+	if op.states.IsListingFullyStored(info.listingID) {
+		// lock map is shared with the collector, locked there
+		// we unlock when all the object were stored (ie: acked)
+		lock, _ := op.lockMap.Load(info.listingID)
+		lock.(*sync.Mutex).Unlock()
+	}
+}
+
+type listingInfo struct {
+	totObjects    int
+	storedObjects int
+}
+
+// States handles list of s3 object state. One must use NewStates to instantiate a
+// file states registry. Using the zero-value is not safe.
+type States struct {
```

Comment:
Does this need to be exported from the package? If there are no external users then I'd make it unexported. Same for `State`.

Reply:
👍